### PR TITLE
Dev: ui_corosync: Improve corosync.show and corosync.edit subcommands

### PR DESCRIPTION
**Pull request by liangxin1300**
- corosync.show: validate corosync.conf before showing it
- corosync.edit: edit a temporary copy of corosync.conf, validate it before saving it,
    then give the user hints on how to sync to other nodes
- Add corosync_ui.feature to collect test cases for crm corosync ui

### CODE REVIEW

**Comment by liangxin1300** [comment]
- [ ] How to loosen the limitation of codecov to allow such interactive cases which have 0% diff hit?

File: crmsh/utils.py [hunk 1]
```diff
@@ -1221,31 +1221,22 @@
...
-            f2 = open(fname, 'w')
-            f2.write(s)
-            f2.close()
-        finally:
-            os.unlink(tmpfile)
-    except OSError as e:
-        raise IOError(e)
+    with create_tempfile() as tmpfile:
+        source_data = read_from_file(fname)
+        str2file(source_data, tmpfile)
```
Comment (nicholasyang2022) [inline crmsh/utils.py:1232]: `str2file` uses `open_atomic`. It does not make sense to it for writing a tempfile. Something like `shutil.copyfile` is enough.

File: crmsh/utils.py [hunk 2]
```diff
@@ -1221,31 +1221,22 @@ def edit_file(fname):
     return ext_cmd_nosudo("%s %s" % (config.core.editor, fname))
 
 
-def edit_file_ext(fname, template=''):
+@contextmanager
+def edit_file_ext(fname):
```
Comment (nicholasyang2022) [inline crmsh/utils.py:1225]: The interface looks strange. `edit_file_ext` does not finished the job of editing a file. How about `def edit_file_ext(fname: str, validator: typing.Callable[typing.IO, bool])`

File: crmsh/ui_corosync.py [hunk 3]
```diff
@@ -121,18 +121,29 @@ def do_edit(self, context):
...
         '''
         Display the corosync configuration.
         '''
+        if not corosync.is_valid_corosync_conf():
+            return False
         cfg = corosync.conf()
-        if not os.path.isfile(cfg):
-            context.fatal_error("No corosync configuration found on this node.")
-        utils.page_string(open(cfg).read())
+        utils.page_string(utils.read_from_file(cfg))
```
Comment (nicholasyang2022) [inline crmsh/ui_corosync.py:146]: No need to read it into memory and send to pager. `page_file` is enough.

File: crmsh/utils.py [hunk 4]
```diff
@@ -1221,31 +1221,26 @@ def edit_file(fname):
...
+    with create_tempfile() as tmpfile:
+        shutil.copyfile(fname, tmpfile)
+        if edit_file(tmpfile) != 0:
+            raise IOError(f"Cannot edit file \"{fname}\"")
+        changed_data = read_from_file(tmpfile)
+        source_data = read_from_file(fname)
+        if hash(changed_data) != hash(source_data):
+            if validator and not validator(tmpfile):
+                return False
+            shutil.copyfile(tmpfile, fname)
```
Comment (nicholasyang2022) [inline crmsh/utils.py:1240]: The original file needs to be replaced atomically.  Either create the temp file in the same directory, fsync and use `os.rename` to replace the original file, or use `str2file`.

Comment (liangxin1300) [reply crmsh/utils.py]: Changed, please take a look
Thanks!

File: crmsh/utils.py [hunk 5]
```diff
@@ -1230,31 +1230,27 @@ def edit_file(fname):
...
-            os.unlink(tmpfile)
-    except OSError as e:
-        raise IOError(e)
+    with create_tempfile() as tmpfile:
+        shutil.copyfile(fname, tmpfile)
+        if edit_file(tmpfile) != 0:
+            raise IOError(f"Cannot edit file \"{fname}\"")
+        changed_data = read_from_file(tmpfile)
+        source_data = read_from_file(fname)
+        if hash(changed_data) != hash(source_data):
```
Comment (nicholasyang2022) [inline crmsh/utils.py:1246]: `hash(a) != hash(b)` is not a necessary condition for `a != b`.

**Review by nicholasyang2022 — CHANGES_REQUESTED** [review]
1. 2 pieces of data shares a same hash value does not means those data are the same.
2. Please update `after_n_builds` in `codecov.yml` as a new test job is added.

File: crmsh/utils.py (hunk 5 above)
Comment (liangxin1300) [reply crmsh/utils.py]: Changed as comparing them directly

**Review by nicholasyang2022 — APPROVED** [review]

**Pull request merged**